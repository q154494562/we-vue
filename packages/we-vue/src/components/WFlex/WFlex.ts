import Vue from 'vue'
import '@/scss/flex.scss'
import { PropValidator } from 'vue/types/options'
import { VNode } from 'vue/types/vnode'

export default Vue.extend({
  name: 'w-flex',

  provide (): object {
    return { flexComponent: this }
  },

  props: {
    gutter: {
      type: [Number, String],
      default: 0,
      validator: (val: number | string) => {
        return Number(val) >= 0
      },
    } as PropValidator<number | string>,
  },

  computed: {
    style (): object {
      const margin = `-${Number(this.gutter) / 2}px`

      return this.gutter ? { marginLeft: margin, marginRight: margin } : {}
    },
  },

  render (h): VNode {
    return h('div', {
      class: 'weui-flex',
      style: this.style,
    }, [
      this.$slots.default,
    ])
  },
})
