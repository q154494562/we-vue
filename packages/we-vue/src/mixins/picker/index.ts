// Mixins
import Toggleable, { factory as ToggleableFactory } from '../toggleable'

// Utils
import mixins, { ExtractVue } from '@/utils/mixins'

export default mixins<
  ExtractVue<[typeof Toggleable]>
>(
  ToggleableFactory('visible', 'update:visible')
).extend({
  name: 'picker',

  props: {
    confirmText: {
      type: String,
      default: '确定',
    },
    cancelText: {
      type: String,
      default: '取消',
    },
    closeOnClickMask: {
      type: Boolean,
      default: true,
    },
  },

  methods: {
    open (): void {
      this.isActive = true
    },

    close (): void {
      this.isActive = false
    },
  },
})
