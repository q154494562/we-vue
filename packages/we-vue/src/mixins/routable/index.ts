import Vue from 'vue'

export default Vue.extend({
  name: 'routable',

  props: {
    href: String,
    replace: Boolean,
    to: [String, Object],
  },

  methods: {
    routeLink () {
      const { to, href, $router, replace } = this as any
      if (to && $router) {
        $router[replace ? 'replace' : 'push'](to)
      } else if (href) {
        replace ? location.replace(href) : location.href = href
      }
    },
  },
})
