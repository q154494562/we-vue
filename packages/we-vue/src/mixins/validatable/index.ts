import Vue from 'vue'

export default Vue.extend({
  name: 'validatable',

  props: {
    value: {
      required: false,
    },
  },

  data () {
    return {
      lazyValue: this.value,
      valid: false,
    }
  },
})
