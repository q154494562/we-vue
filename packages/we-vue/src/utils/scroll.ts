type ElementOrWindow = HTMLElement | Window

/**
 * 找到最近的触发滚动事件的元素
 * @param {Element} element
 * @param {Element} rootParent
 * @returns {Element | window}
 */
export function getScrollEventTarget (element: HTMLElement, rootParent: ElementOrWindow = window) {
  let node = element
  while (
    node &&
    node.tagName !== 'HTML' &&
    node.tagName !== 'BODY' &&
    node.nodeType === 1 &&
    node !== rootParent
  ) {
    const { overflowY } = window.getComputedStyle(node)
    if (overflowY === 'scroll' || overflowY === 'auto') {
      return node
    }
    node = <HTMLElement>node.parentNode
  }
  return rootParent
}

// 获取滚动高度
export function getScrollTop (element: ElementOrWindow): number {
  return 'scrollTop' in element ? element.scrollTop : element.pageYOffset
}

// 设置滚动高度
export function setScrollTop (element: ElementOrWindow, value: number) {
  'scrollTop' in element ? (element.scrollTop = value) : element.scrollTo(element.scrollX, value)
}

// 获取元素距离顶部高度
export function getElementTop (element: ElementOrWindow) {
  return (
    (element === window ? 0 : (element as HTMLElement).getBoundingClientRect().top) + getScrollTop(window)
  )
}

export function getVisibleHeight (element: ElementOrWindow) {
  return element === window
    ? element.innerHeight
    : (element as HTMLElement).getBoundingClientRect().height
}
