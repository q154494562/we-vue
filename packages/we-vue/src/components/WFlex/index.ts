import WFlex from './WFlex'

export { WFlex }
export default WFlex
