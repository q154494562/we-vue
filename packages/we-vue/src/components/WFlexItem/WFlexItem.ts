import Vue, { VNode } from 'vue'

import WFlex from '../WFlex'

type WFlexInstance = InstanceType<typeof WFlex>

interface options extends Vue {
  flexComponent: WFlexInstance
}

export default Vue.extend<options>().extend({
  name: 'w-flex-item',

  inject: {
    flexComponent: {
      default: null,
    },
  },

  props: {
    flex: {
      type: [Number, String],
      default: 1,
    },
    offset: {
      type: String,
      default: '',
    },
  },

  computed: {
    gutter (): number {
      return (Number(this.flexComponent.gutter)) || 0
    },

    style (): object {
      const padding = `${Number(this.gutter) / 2}px`

      const ret = this.gutter
        ? {
          paddingLeft: padding,
          paddingRight: padding,
        }
        : {}

      return { ...ret, ...{ flex: Number(this.flex), marginLeft: this.offset } }
    },
  },

  render (h): VNode {
    return h('div', {
      class: 'weui-flex__item',
      style: this.style,
    }, [
      this.$slots.default,
    ])
  },
})
