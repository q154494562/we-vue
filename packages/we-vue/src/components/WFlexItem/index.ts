import WFlexItem from './WFlexItem'

export { WFlexItem }
export default WFlexItem
