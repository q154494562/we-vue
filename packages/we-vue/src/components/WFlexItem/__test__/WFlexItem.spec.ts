import { mount, shallowMount, Wrapper } from '@vue/test-utils'
import WFlex from '../../WFlex'
import WFlexItem from '../'
import { ExtractVue } from '@/utils/mixins'

describe('flex-item', () => {
  type FlexWrapper = Wrapper<ExtractVue<typeof WFlex>>
  type FlexItemWrapper = Wrapper<ExtractVue<typeof WFlexItem>>

  test('create', () => {
    const parentWrapper = mount(WFlex, {
      propsData: {
        animate: true,
      },
    })

    const wrapper = shallowMount(WFlexItem, {
      provide: {
        flexComponent: parentWrapper.vm,
      },
    })

    expect(wrapper.name()).toBe('w-flex-item')
    expect(wrapper.classes()).toContain('weui-flex__item')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('comput gutter', () => {
    const wrapper = mount(WFlex, {
      attachToDocument: true,
      propsData: {
        gutter: 10,
      },
      slots: {
        default: WFlexItem,
      },
    }) as FlexWrapper

    const _flexItem = wrapper.find(WFlexItem) as FlexItemWrapper

    expect(_flexItem.vm.gutter).toBe(wrapper.vm.gutter)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('comput style', () => {
    const parentWrapper = mount(WFlex, {
      propsData: {},
    })

    const wrapper = mount(WFlexItem, {
      provide: {
        flexComponent: parentWrapper.vm,
      },
    }) as Wrapper<ExtractVue<[typeof WFlexItem]>>

    const computedStyle: {[key: string]: any} = {
      flex: wrapper.vm.flex,
      marginLeft: wrapper.vm.offset,
    }

    if (wrapper.vm.gutter) {
      computedStyle.paddingLeft = wrapper.vm.gutter / 2 + 'px'
      computedStyle.paddingRight = computedStyle.paddingLeft
    }

    expect(wrapper.vm.style).toEqual(computedStyle)
  })
})
