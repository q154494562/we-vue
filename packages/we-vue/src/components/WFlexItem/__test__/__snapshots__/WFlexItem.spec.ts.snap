// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flex-item comput gutter 1`] = `
<div class="weui-flex" style="margin-left: -5px; margin-right: -5px;">
  <div class="weui-flex__item" style="padding-left: 5px; padding-right: 5px; flex: 1;"></div>
</div>
`;

exports[`flex-item create 1`] = `<div class="weui-flex__item" style="flex: 1;"></div>`;
